classes/class-p4-custom-taxonomy.php

```diff
@@ -0,0 +1,233 @@
+<?php
+
+if ( ! class_exists( 'P4_Custom_Taxonomy' ) ) {
+
+	/**
+	 * Class P4_Custom_Taxonomy
+	 */
+	class P4_Custom_Taxonomy {
+
+		const TAXONOMY = 'p4-page-type';
+
+		/**
+		 * P4_Custom_Taxonomy constructor.
+		 */
+		public function __construct() {
+			$this->hooks();
+		}
+
+		/**
+		 * Register actions for WordPress hooks and filters.
+		 */
+		private function hooks() {
+			add_action( 'init',                               array( $this, 'register_taxonomy' ), 2 );
+			add_action( 'created_term',                       array( $this, 'trigger_rewrite_rules' ), 10, 3 );
+			add_action( 'edited_term',                        array( $this, 'trigger_rewrite_rules' ), 10, 3 );
+			add_action( 'delete_term',                        array( $this, 'trigger_rewrite_rules' ), 10, 3 );
+			add_action( 'save_post',                          array( $this, 'save_taxonomy_page_type' ) );
+			add_filter( 'available_permalink_structure_tags', array( $this, 'add_taxonomy_as_permalink_structure' ), 10, 1 );
+			add_filter( 'post_link',                          array( $this, 'filter_permalink' ), 10, 3 );
+			add_filter( 'post_rewrite_rules',                 array( $this, 'replace_taxonomy_terms_in_rewrite_rules' ), 10, 1 );
+		}
+
+		/**
+		 * Add p4_page_type structure in available permalink tags for Settings -> Permalinks page.
+		 * available_permalink_structure_tags filter.
+		 *
+		 * @param array $tags   Permalink tags that are displayed in Settings -> Permalinks.
+		 *
+		 * @return mixed
+		 */
+		public function add_taxonomy_as_permalink_structure( $tags ) {
+			$tags['p4_page_type'] = __( 'P4 page type (A p4 page type term.)', 'planet4-master-theme' );
+
+			return $tags;
+		}
+
+		/**
+		 * Add a dropdown to choose planet4 post type.
+		 *
+		 * @param WP_Post $post The WordPress that will be filtered/edited.
+		 */
+		public function create_taxonomy_metabox_markup( WP_Post $post ) {
+			$attached_type = get_the_terms( $post, self::TAXONOMY );
+			$current_type  = ( is_array( $attached_type ) ) ? $attached_type[0]->slug : - 1;
+			$all_types     = get_terms( self::TAXONOMY, [ 'hide_empty' => false ] );
+			wp_nonce_field( 'p4-save-page-type', 'p4-page-type-nonce' );
+			?>
+			<select name="<?php echo self::TAXONOMY; ?>" disabled>
```

Comment:
Could you escape the name attribute please. I know it is a constant that you output but still please 
`echo esc_attr( self::TAXONOMY );` just to stop CS from complaining about standards.